### PR TITLE
Refactor response to action in agent step

### PR DESCRIPTION
**Pull request by yufansong**
With the improvement of our agent, more and more actions will add into agent. The logic of converting LLM response to action become complicated.
 
In each step: 
1. We need to call LLM get response
2. Parse the response into action str
3. Convert the action string into the actual `Action`. It may involve lots of actions.

So I create two abstract class:
1. `ResponseParser`: take over whole logic of convert LLM response to action
2. `ActionParser`: convert action string into one specific action.

Currently I only refactor the `CodeactAgent` which are most complicated in agenthubs. If others agree with this refactor, I will refactor other agents in future PR.

### CODE REVIEW

File: agenthub/codeact_agent/action_parser.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    Action,
+    AgentDelegateAction,
+    AgentFinishAction,
+    CmdRunAction,
+    IPythonRunCellAction,
+    MessageAction,
+)
+
+
+class CodeactResponseParser(ResponseParser):
```
Comment (tobitege) [inline agenthub/codeact_agent/action_parser.py:14]: maybe name as `CodeActResponseParser` instead?

**Review by li-boxuan — APPROVED** [review]
LGTM 👍 

Agree with @tobitege that `Codeact` should be `CodeAct`

File: agenthub/codeact_agent/action_parser.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    """
+
+    def __init__(
+        self,
+    ):
+        pass
+
+    def check_condition(self, action_str: str) -> bool:
+        # We assume the LLM is GOOD enough that when it returns pure natural language
+        # it want to talk to the user
```
Comment (li-boxuan) [inline agenthub/codeact_agent/action_parser.py:178]:
```suggestion
        # it wants to talk to the user
```

File: opendevin/controller/action_parser.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+from abc import ABC, abstractmethod
+
+from opendevin.events.action import Action
+
+
+class ResponseParser(ABC):
+    """
+    This abstract base class is an general interface for an response parser dedicated to
```
Comment (li-boxuan) [inline opendevin/controller/action_parser.py:8]:
```suggestion
    This abstract base class is a general interface for an response parser dedicated to
```

**Pull request merged**